src/classes/CENR_AcademicProgram_TDTM.cls

```diff
@@ -35,69 +35,142 @@
+                Course_Enrollment__c enroll = (Course_Enrollment__c)newCourseEnrollments[i];
+                //Get the ID of each parent Program Enrollment
+                if (enroll.Program_Enrollment__c != null) {
+                    pEnrollIDsToQuery.add(enroll.Program_Enrollment__c);
                 }
             }
-            
-            //Query the parent Account (Academic Program) of each Program Enrollment, since that field
-            //wasn't in the trigger for being two levels up.
-            List<Program_Enrollment__c> pEnrollments = [select Account__c from Program_Enrollment__c 
-                                                    where Account__c != null and ID in :pEnrollIDsToQuery];
-            Map<Id, Program_Enrollment__c> pEnrollmentsMap = new Map<Id, Program_Enrollment__c>(pEnrollments); 
-            
-            for (Integer i = 0; i < newlist.size(); i++) {
-                Course_Enrollment__c enroll = (Course_Enrollment__c)newlist[i];
-                
+
+            if (pEnrollIDsToQuery.size() > 0) {
+                pEnrollments = queryProgramEnrollments(pEnrollIDsToQuery);
+            }
+
+            for (Integer i = 0; i < newCourseEnrollments.size(); i++) {
+                Map<Id, Program_Enrollment__c> pEnrollmentsMap = new Map<Id, Program_Enrollment__c>(pEnrollments);
+                Course_Enrollment__c enroll = (Course_Enrollment__c)newCourseEnrollments[i];
+
                 //BEFORE INSERT
-                if(triggerAction == TDTM_Runnable.Action.BeforeInsert) {
-                    //set Academic Program if we can get it through Program Enrollment
-                    if(enroll.Program_Enrollment__c != null) {
-                        Program_Enrollment__c pEnroll = pEnrollmentsMap.get(enroll.Program_Enrollment__c);
-                        if(pEnroll != null) //shouldn't really happen... we already queried for this...
-                            enroll.Account__c = pEnroll.Account__c; 
-                    }
+                if (triggerAction == TDTM_Runnable.Action.BeforeInsert) {
+                    processBeforeInsert(enroll, pEnrollmentsMap, affiliationIsEnabled);
                 }
-                
+
                 //BEFORE UPDATE
-                if(triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
-                    Course_Enrollment__c oldEnroll = (Course_Enrollment__c)oldList[i];
-                    //set Academic Program if we can get it through Program Enrollment
-                    if(oldEnroll.Program_Enrollment__c == null && enroll.Program_Enrollment__c != null) {
-                        Program_Enrollment__c pEnroll = pEnrollmentsMap.get(enroll.Program_Enrollment__c);
-                        if(pEnroll != null) //shouldn't really happen... we already queried for this...
-                            enroll.Account__c = pEnroll.Account__c;
-                    }
+                if (triggerAction == TDTM_Runnable.Action.BeforeUpdate) {
```

Comment:
@lifewithryan , unfortunately I don't think we can leverage oldMap/newMap since the main run method's signature is accepting list and not maps. 

As for the request to consolidate the 2 processBeforeInsert && processBeforeUpdate methods, those are now done and have been replaced with just 1 method called assignAccountAndAffiliation() method.